File: hdf-converters/shared/go/amendmentchain_test.go
package shared

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
	"time"

	hdf "github.com/mitre/hdf-libs/hdf-schema/dist/go/v3"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func chainWaiver(id, reason string) hdf.StandaloneOverride {
	return hdf.StandaloneOverride{
		Type:          hdf.OverrideType("waiver"),
		RequirementID: id,
		Reason:        reason,
		AppliedBy:     hdf.Identity{Type: hdf.Email, Identifier: "admin@example.com"},
		AppliedAt:     time.Date(2026, 3, 1, 12, 0, 0, 0, time.UTC),
		ExpiresAt:     time.Date(2099, 12, 31, 0, 0, 0, 0, time.UTC),
	}
}

func TestChainOverrides(t *testing.T) {
	t.Run("leaves the first unlinked and chains the rest", func(t *testing.T) {
		overrides := []hdf.StandaloneOverride{
			chainWaiver("AC-1", "first"), chainWaiver("AC-2", "second"), chainWaiver("AC-3", "third"),
		}
		require.NoError(t, ChainOverrides(overrides))

		assert.Nil(t, overrides[0].PreviousChecksum)
		require.NotNil(t, overrides[1].PreviousChecksum)
		assert.Equal(t, hdf.Sha256, overrides[1].PreviousChecksum.Algorithm)
		assert.Len(t, overrides[1].PreviousChecksum.Value, 64)
		require.NotNil(t, overrides[2].PreviousChecksum)
		assert.NotEqual(t, overrides[1].PreviousChecksum.Value, overrides[2].PreviousChecksum.Value)
	})

	t.Run("is deterministic", func(t *testing.T) {
		a := []hdf.StandaloneOverride{chainWaiver("AC-1", "x"), chainWaiver("AC-2", "y")}
		b := []hdf.StandaloneOverride{chainWaiver("AC-1", "x"), chainWaiver("AC-2", "y")}
		require.NoError(t, ChainOverrides(a))
		require.NoError(t, ChainOverrides(b))
		assert.Equal(t, a[1].PreviousChecksum.Value, b[1].PreviousChecksum.Value)
	})

	t.Run("the link changes when the preceding override changes", func(t *testing.T) {
		before := []hdf.StandaloneOverride{chainWaiver("AC-1", "original"), chainWaiver("AC-2", "second")}
		after := []hdf.StandaloneOverride{chainWaiver("AC-1", "edited"), chainWaiver("AC-2", "second")}
		require.NoError(t, ChainOverrides(before))
		require.NoError(t, ChainOverrides(after))
		assert.NotEqual(t, before[1].PreviousChecksum.Value, after[1].PreviousChecksum.Value)
	})

	t.Run("re-chaining an already-chained set is stable", func(t *testing.T) {
		overrides := []hdf.StandaloneOverride{chainWaiver("AC-1", "first"), chainWaiver("AC-2", "second")}
		require.NoError(t, ChainOverrides(overrides))
		first := overrides[1].PreviousChecksum.Value
		require.NoError(t, ChainOverrides(overrides))
		assert.Equal(t, first, overrides[1].PreviousChecksum.Value)
	})

	t.Run("an empty set is a no-op", func(t *testing.T) {
		assert.NoError(t, ChainOverrides(nil))
	})

	// The TypeScript chainer asserts against this same fixture, so the two
	// languages are pinned to one chain rather than to each other's promises.
	t.Run("reproduces the checksums in the expected fixture", func(t *testing.T) {
		path := filepath.Join("..", "..", "converters", "openvex-to-hdf", "fixtures",
			"expected", "multi-status.openvex.json.hdf.json")
		raw, err := os.ReadFile(path) //nolint:gosec // fixture path is repo-relative
		require.NoError(t, err)

		var doc hdf.HDFAmendments
		require.NoError(t, json.Unmarshal(raw, &doc))
		require.GreaterOrEqual(t, len(doc.Overrides), 2)

		recorded := make([]*hdf.Checksum, len(doc.Overrides))
		for i := range doc.Overrides {
			recorded[i] = doc.Overrides[i].PreviousChecksum
		}

		require.NoError(t, ChainOverrides(doc.Overrides))
		for i := range doc.Overrides {
			if recorded[i] == nil {
				assert.Nilf(t, doc.Overrides[i].PreviousChecksum, "override %d", i)
				continue
			}
			require.NotNilf(t, doc.Overrides[i].PreviousChecksum, "override %d", i)
			assert.Equalf(t, recorded[i].Value, doc.Overrides[i].PreviousChecksum.Value, "override %d", i)
		}
	})
}

File: hdf-converters/shared/typescript/amendmentchain.test.ts
import { describe, expect, it } from 'vitest';
import { readFileSync } from 'node:fs';
import { HashAlgorithm } from '@mitre/hdf-schema';
import type { StandaloneOverride } from '@mitre/hdf-schema';
import { chainOverrides } from './amendmentchain.js';

function waiver(id: string, reason: string): StandaloneOverride {
  return {
    type: 'waiver',
    requirementId: id,
    status: 'passed',
    reason,
    appliedBy: { type: 'email', identifier: 'admin@example.com' },
    appliedAt: new Date('2026-03-01T12:00:00Z'),
    expiresAt: new Date('2099-12-31T00:00:00Z'),
  } as unknown as StandaloneOverride;
}

describe('chainOverrides', () => {
  it('leaves the first override unlinked and chains the rest', async () => {
    const overrides = [waiver('AC-1', 'first'), waiver('AC-2', 'second'), waiver('AC-3', 'third')];
    await chainOverrides(overrides);

    expect(overrides[0].previousChecksum).toBeUndefined();
    expect(overrides[1].previousChecksum?.algorithm).toBe(HashAlgorithm.Sha256);
    expect(overrides[1].previousChecksum?.value).toMatch(/^[0-9a-f]{64}$/);
    expect(overrides[2].previousChecksum?.value).not.toBe(overrides[1].previousChecksum?.value);
  });

  it('is deterministic for identical input', async () => {
    const a = [waiver('AC-1', 'x'), waiver('AC-2', 'y')];
    const b = [waiver('AC-1', 'x'), waiver('AC-2', 'y')];
    await chainOverrides(a);
    await chainOverrides(b);
    expect(a[1].previousChecksum?.value).toBe(b[1].previousChecksum?.value);
  });

  it('changes the recorded link when the preceding override changes', async () => {
    const before = [waiver('AC-1', 'original'), waiver('AC-2', 'second')];
    const after = [waiver('AC-1', 'edited'), waiver('AC-2', 'second')];
    await chainOverrides(before);
    await chainOverrides(after);
    expect(after[1].previousChecksum?.value).not.toBe(before[1].previousChecksum?.value);
  });

  it('re-chaining an already-chained set is stable', async () => {
    const overrides = [waiver('AC-1', 'first'), waiver('AC-2', 'second')];
    await chainOverrides(overrides);
    const first = overrides[1].previousChecksum?.value;
    await chainOverrides(overrides);
    expect(overrides[1].previousChecksum?.value).toBe(first);
  });

  // The chain a TS converter writes must verify under Go's `hdf amend verify`.
  // The Go-generated expected fixtures are the contract; this asserts the link
  // this code produces is the one those fixtures record.
  it('reproduces the checksums in the Go-generated expected fixture', async () => {
    const expected = JSON.parse(
      readFileSync(
        'converters/openvex-to-hdf/fixtures/expected/multi-status.openvex.json.hdf.json',
        'utf8',
      ),
    ) as { overrides: StandaloneOverride[] };

    const recorded = expected.overrides.map((o) => o.previousChecksum?.value);
    const rechained = expected.overrides.map((o) => ({ ...o }));
    await chainOverrides(rechained as StandaloneOverride[]);

    expect(rechained.map((o) => o.previousChecksum?.value)).toEqual(recorded);
    expect(recorded.filter(Boolean).length).toBeGreaterThan(0);
  });
});
